Content.Server/Supermatter/EntitySystems/SupermatterSystem.cs

```diff
@@ -0,0 +1,425 @@
+                _damageable.TryChangeDamage(SMcomponent.Owner, damageDelta, true);
+            }
+        }
+
+        /// </summary>
+        /// Runs the logic and timers for Delamination
+        /// </summary>
+        public void Delamination(EntityUid Uid, float frameTime, SupermatterComponent? SMcomponent = null, TransformComponent? Xform = null)
+        {
+            if (!Resolve(Uid, ref SMcomponent, ref Xform))
+            {
+                return;
+            }
+
+            //TODO: make tesla spawn at SupermatterComponent.PowerPenaltyThreshold
+            if (!SMcomponent.FinalCountdown)
+            {
+                if (_atmosphereSystem.GetTileMixture(Xform.Coordinates) is { } mixture)
+                {
+                    if (mixture.TotalMoles >= SupermatterComponent.MolePenaltyThreshold)
+                    {
+                        _chatManager.DispatchStationAnnouncement(Loc.GetString("supermatter-delamination-overmass"), "Supermatter", false);
+                    }
+                    else
+                    {
+                        _chatManager.DispatchStationAnnouncement(Loc.GetString("supermatter-delamination-default"), "Supermatter", false);
+                    }
+                }
+            }
+            SMcomponent.FinalCountdown = true;
+
+            SMcomponent.DelamTimerAccumulator += frameTime;
+            SMcomponent.SpeakAccumulator += frameTime;
+            int RoundSeconds = SupermatterComponent.DelamTimerTimer - (int)Math.Floor(SMcomponent.DelamTimerAccumulator);
+
+            if (SMcomponent.DamageArchived < SupermatterComponent.ExplosionPoint)
+            {
+                _chatManager.DispatchStationAnnouncement(Loc.GetString("supermatter-safe-allert"), "Supermatter", false);
+                SMcomponent.FinalCountdown = false;
+                return;
+            }
+            else if (RoundSeconds >= 5 && SMcomponent.SpeakAccumulator >= 5)
+            {
+                SMcomponent.SpeakAccumulator -= 5;
+                _chatManager.DispatchStationAnnouncement(Loc.GetString("supermatter-seconds-before-delam", ("Seconds", RoundSeconds)), "Supermatter", false);
+            }
+            else if (RoundSeconds <  5 && SMcomponent.SpeakAccumulator >= 1)
+            {
+                SMcomponent.SpeakAccumulator -= 1;
+                _chatManager.DispatchStationAnnouncement(Loc.GetString("supermatter-seconds-before-delam", ("Seconds", RoundSeconds)), "Supermatter", false);
+            }
+
+            if (SMcomponent.DelamTimerAccumulator >= SupermatterComponent.DelamTimerTimer)
+            {
+                if (_atmosphereSystem.GetTileMixture(Xform.Coordinates) is { } mixture)
+                {
+                    if (mixture.TotalMoles >= SupermatterComponent.MolePenaltyThreshold)
+                    {
+                        EntityManager.SpawnEntity("Singularity", Xform.Coordinates);
```

Comment:
Make this a field on the SM or something so we at least have some kind of verification it works in CI.